gobblin-modules/gobblin-kafka-common/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -591,11 +602,33 @@ private void startAtLatestOffset() {
     @Override
     public void run() {
       try (Timer.Context context = metricContext.timer(WORK_UNITS_FOR_TOPIC_TIMER).time()) {
+        // use shared client if configure, otherwise create a thread local one
+        if (KafkaSource.this.sharedKafkaConsumerClient != null) {
+          KafkaSource.this.kafkaConsumerClient.set(KafkaSource.this.sharedKafkaConsumerClient);
+        } else {
+          KafkaSource.this.kafkaConsumerClient.set(
```

Comment:
The run() method sets the ThreadLocal using KafkaSource.this.kafkaConsumerClient.set(). This sets one kafkaConsumerClient per thread, so that the references to kafkaConsumerClient elsewhere in KafkaSource, such as getWorkunits() will use the one for allocated for the thread.